File: Rabin-Karp/Java/RabinKarp.java
/*
 * Author:- Prarik Kayastha
 * Email-id:- pratikkayastha98@gmail.com
 * Program Name:- Rabin-Karp Algorithm
 * Description:- This algorithm uses to find pattern in given string.
 * Time-Complexity:- O(mn)
 * */

import java.util.Scanner;

public class RabinKarp {

	static final long prime = 101;
	public static String searchSubstring(String str,int n,String sub,int m)
	{
		long key= getSubKey(sub, m);
		long oldHash = getSubKey(str.substring(0, m), m);
		if(key==oldHash && equal(str, sub, 0))
			return "Yes";
		for(int i=m;i<n;i++)
		{
			oldHash = getNewHash(str, i-m, i, oldHash, m);
			if(key==oldHash && equal(str, sub, i-m+1))
				return "Yes";
		}
		return "No";
	}
	public static long getNewHash(String str,int oldIndex,int newIndex,long oldHash,int m)//get newhash in constant time
	{
		long newHash=((oldHash-str.charAt(oldIndex)+96)/prime)+((str.charAt(newIndex)-96)*(long)Math.pow(prime, m-1));
		return newHash;
	}
	public static long getSubKey(String sub,int m)//hashing function
	{
		long key=0;
		for(int i=0;i<m;i++)
		{
			key += (sub.charAt(i)-96)*(long)Math.pow(prime, i);
		}
		return key;
	}
	public static boolean equal(String str,String sub,int index)//to check two string are equal or not
	{
		for(int i=0;i<sub.length();i++)
			if(str.charAt(index+i)!=sub.charAt(i))
				return false;
		return true;
	}
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Scanner sc = new Scanner(System.in);
		String str = sc.nextLine();
		String sub = sc.nextLine();
		int n = str.length();
		int m = sub.length();
		System.out.println(searchSubstring(str,n,sub,m));
		sc.close();
	}

}

File: RobinCarp/Kotlin/RabinKarp.kt
val prime: Long = 101
fun searchSubstring(str: String, n: Int, sub: String, m: Int): String {
    val key = getSubKey(sub, m)
    var oldHash = getSubKey(str.substring(0, m), m)
    if (key == oldHash && equal(str, sub, 0))
        return "Yes"
    for (i in m until n) {
        oldHash = getNewHash(str, i - m, i, oldHash, m)
        if (key == oldHash && equal(str, sub, i - m + 1))
            return "Yes"
    }
    return "No"
}

fun getNewHash(str: String, oldIndex: Int, newIndex: Int, oldHash: Long, m: Int)
        : Long {
    return (oldHash - str[oldIndex].toLong() + 96) / prime + (str[newIndex].toInt() - 96) * Math.pow(prime.toDouble(), (m - 1).toDouble()).toLong()
}

fun getSubKey(sub: String, m: Int)
        : Long {
    var key: Long = 0
    for (i in 0 until m) {
        key += (sub[i].toInt() - 96) * Math.pow(prime.toDouble(), i.toDouble()).toLong()
    }
    return key
}

fun equal(str: String, sub: String, index: Int)//to check two string are equal or not
        : Boolean {
    for (i in 0 until sub.length)
        if (str[index + i] != sub[i])
            return false
    return true
}


fun main(args: Array<String>) {

    
    val str = readLine()!!
    val sub = readLine()!!
    val n = str.length
    val m = sub.length
    println(searchSubstring(str, n, sub, m))
    
}
